File: java/37_SudokuSolver.java
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        def valid_candidate(cand, row, col):
            for i in range(9):
                if row != i and board[i][col] == cand: return False
                if col != i and board[row][i] == cand: return False
                
                box_row = row // 3 * 3 + i // 3
                box_col = col // 3 * 3 + i % 3
                if (row != box_row and col != box_col) and board[box_row][box_col] == cand: return False
            return True
        
        def solve():
            for i in range(9):
                for j in range(9):
                    if board[i][j] == '.':
                        for n in range(1, 10):
                            if valid_candidate(str(n), i, j):
                                board[i][j] = str(n)
                                if solve(): return True
                                board[i][j] = '.'
                        return False
            return True
        
        solve()

File: cpp/37_SudokuSolver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        backtrack(board);
    }
    
    bool checkEntry(vector<vector<char>>& board, char cand, int row, int col) {
        int box_row = row / 3 * 3, box_col = col / 3 * 3;
        for (int i = 0; i < 9; ++i) {
            if (row != i && board[i][col] == cand) return false;
            if (col != i && board[row][i] == cand) return false;
            if ((box_row != row && box_col != col) && board[box_row][box_col] == cand) return false;
            
            box_row = row / 3 * 3 + i / 3;
            box_col = col / 3 * 3 + i % 3;
        }
        return true;
    }
    
    bool backtrack(vector<vector<char>>& board) {
        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.') {
                    for (int n = 1; n <= 9; ++n) {
                        char cand = '0' + n;
                        if (checkEntry(board, cand, i, j)) {
                            board[i][j] = cand;
                            if (backtrack(board)) return true;
                            board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
};
